### PR TITLE
Spam2: system for flagging nodes and comments to moderators and User moderation

### PR DESCRIPTION
**Pull request by keshavsethi**
I have added the feature of Flag Nodes in spam2 in which any user can mark node as a flag. It will not affect the status of the node.

Flag will just increase the flag count by 1. If flag is 0 then it means that the node is not flagged and it is ok. All the flagged nodes will be listed on the flags page of spam2. Right now I have done it for nodes and working for flag comments as well.(refer https://github.com/publiclab/plots2/pull/8107#issuecomment-653942167)

If  flag count is above 10 then flag icon also becomes red. 

Here are a few screenshots

Screenshot from 2020-07-02 20-20-22
Screenshot from 2020-07-02 20-21-03

Here flag count in shown in the status column and red icon indicates that the flag count is more than 10. Spamming that node will also unflag it as it is moderated.
Screenshot from 2020-07-02 20-21-05
 
@jywarren @cesswairimu @VladimirMikulic @ananyaarun @emilyashley @pydevsg @ebarry @Uzay-G Please review
Thanks!!

### CODE REVIEW

**Comment by ebarry** [comment]
This is a great idea!
Who sees the red flag, and where? 

- Only moderators looking at the /spam dashboard? 
- Regular users browsing content?

**Comment by keshavsethi** [comment]
> This is a great idea!
> Who sees the red flag, and where?
> 
> * Only moderators looking at the /spam dashboard?
> * Regular users browsing content?

Regular users can just flag a node. List of flag nodes and that red flag icon is only visible to moderators at /spam2/flags. 

Thanks!!

**Comment by keshavsethi** [comment]
> Since everything is centred, you could as well centre this section with control buttons:
> https://i.imgur.com/YC7Xn5y.png

@VladimirMikulic Sure, I will do that. Thanks!!

**Comment by pydevsg** [comment]
@keshavsethi , the UI is pretty good, have you made this response as well or it's currently made for the Desktop version?

**Comment by keshavsethi** [comment]
@jywarren @cesswairimu @VladimirMikulic @ananyaarun @emilyashley @pydevsg @ebarry @Uzay-G 
I have also done flag comments in this PR. Here, any user can mark comment as a flag(it will not affect the status of a comment). Moderators can see those flag comments in /spam2/comments that have all the features like publish,spam, etc. flag is also shown and moderator can identify flag comments through the icon.
It is sorted as per flag count. I am currently working on its bulk moderation and tests
Screenshot_2020-07-06 🎈 Public Lab _spam_comments(3)
Screenshot_2020-07-06 🎈 Public Lab _spam_comments
@jywarren @cesswairimu @VladimirMikulic @ananyaarun @emilyashley @pydevsg @ebarry @Uzay-G 
Please review
Thanks!!

**Comment by keshavsethi** [comment]
> @keshavsethi , the UI is pretty good, have you made this response as well or it's currently made for the Desktop version?

Yes, all the pages in /spam2 are responsive
Screenshot_2020-07-06 🎈 Public Lab _spam_flags

**Comment by keshavsethi** [comment]
> @keshav234156 can you please share screenshot of the change, thanks

@VladimirMikulic @pydevsg I have aligned bulk nav 
Screenshot from 2020-07-06 03-12-05
Thanks!!

**Comment by pydevsg** [comment]
Ah great work man , this UI is really clean and crisp. Keep up the good work @keshavsethi :tada:

**Comment by jywarren** [comment]
Looking really nice! One suggestion is to use lots of Bootstrap tooltips:

https://getbootstrap.com/docs/4.4/components/tooltips/

to add short explanatory texts over things like the flag icon, without needing a lot of space! What do you think?

File: app/controllers/spam2_controller.rb [hunk 1]
```diff
@@ -135,4 +149,36 @@ def batch_unban
...
+        flash[:notice] = 'Node flagged.'
+        redirect_back fallback_location: root_path
+  end
+
+  def remove_flag_node
+    @node = Node.find params[:id]
+    if @node.flag == 0
+      flash[:notice] = 'Node already unflagged.'
+    else
+      @node.unflag_node
```
[the remark this answers is not in the document]
Comment (keshavsethi) [reply app/controllers/spam2_controller.rb]: I have added an if statement to check only moderators and admins can unflag item 
Thanks

File: app/models/comment.rb [hunk 2]
```diff
@@ -193,6 +193,18 @@ def publish
     self
   end
 
+  def flag_comment
+    self.flag += 1
+    save
+    self
+  end
+
+  def unflag_comment
```
[the remark this answers is not in the document]
Comment (keshavsethi) [reply app/models/comment.rb:202]: But for this, we have to show flag count to users as well. Right now unflag means only moderators can unflag (i.e reduce flag count to 0). Other users can just flag nodes i.e increase its flag count by 1

**Comment by keshavsethi** [comment]
> Looking really nice! One suggestion is to use lots of Bootstrap tooltips:
> 
> https://getbootstrap.com/docs/4.4/components/tooltips/
> 
> to add short explanatory texts over things like the flag icon, without needing a lot of space! What do you think?

@jywarren I have added a few in bulk moderation nav.

**Comment by keshavsethi** [comment]
> Hi @keshavsethi I would also like to request that you add some tests -- to demonstrate how it is supposed to work, and to ensure it's protected against future changes. I know this is a lot, but it is really a critical part of a big feature like this so I appreciate it!
> 
> * unit tests for changes to the models
> * functional tests for controller actions
> * system tests for javascript actions (pressing buttons, complex integrated sequences like several actions chained together)
> 
> Does this make sense? Thank you so much!

@jywarren Sure, I am working on tests. Should I do add them to the same PR?

File: app/models/comment.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (cesswairimu) [reply app/models/comment.rb:202]: gotcha, cool, sounds good...thanks for the clarification

**Comment by keshavsethi** [comment]
@jywarren @cesswairimu can we merge this? 
Thanks!!

**Comment by cesswairimu** [comment]
Hey @keshavsethi this is coming on great :tada: :tada: . If the tests are ready you can add them here then we can get this merged. Thanks

**Comment by keshavsethi** [comment]
> Hey @keshavsethi this is coming on great   . If the tests are ready you can add them here then we can get this merged. Thanks

@cesswairimu @jywarren @pydevsg @VladimirMikulic @ebarry Tests are added for flag posts. Please review.
Thanks!!

**Comment by keshavsethi** [comment]
I have updated the UI and have added alerts after any moderation. Please refer to the following Screenshot
ezgif com-crop (5)
Screenshot from 2020-07-13 09-30-25
@jywarren @cesswairimu @pydevsg Please review, Thanks!!

**Comment by cesswairimu** [comment]
restarting travis

File: test/functional/spam2_controller_test.rb [hunk 3]
```diff
@@ -228,4 +249,70 @@ def teardown
         assert_equal "Only admins and moderators can batch delete.", flash[:error]
         assert_redirected_to "/dashboard"
     end
+    #flag Nodes
+    test 'Anyone user should be able to flag a note' do
```
Comment (cesswairimu) [inline test/functional/spam2_controller_test.rb:253]: Great job writing the tests...lets fix the grammar here 'Any user  should...' or 'Anyone should...'

File: test/functional/spam2_controller_test.rb [hunk 4]
```diff
@@ -228,4 +249,70 @@ def teardown
...
+      
+    test 'Admins should be able to unflag nodes' do
+        UserSession.create(users(:admin))
+        node = nodes(:about)
+        post :remove_flag_node, params: { id: node.id }
+        node = assigns(:node)
+        assert_equal 0, node.flag
+    end
+    #flag Comments
+    test 'Anyone user should be able to flag a comments' do
```
Comment (cesswairimu) [inline test/functional/spam2_controller_test.rb:286]: Same here  'Any user  should...' or 'Anyone should...'

File: test/functional/spam2_controller_test.rb (hunk 3 above)
Comment (keshavsethi) [reply test/functional/spam2_controller_test.rb]: Sorrry :sweat_smile:  for that, Done. Thanks!!

File: test/functional/spam2_controller_test.rb (hunk 4 above)
Comment (keshavsethi) [reply test/functional/spam2_controller_test.rb]: Done Thanks!!

**Comment by jywarren** [comment]
Hi @keshavsethi this is looking tremendous. I wonder if "Flag this node" should be "Flag this to moderators" as that way it's known who sees the flags, and the vocab word "node" isn't as confusing to folks? 

Also, i wanted to see if you had any thoughts on how to complete #8122 because it occurred to me that it would make reviewing big PRs like this much easier; anyone could just boot it up in GitPod right from the PR! @Tlazypanda this could benefit some of your work as well. Would either of you be interested in helping to make progress on #8122 and that could speed up our review time quite a bit! 🎉

File: test/functional/spam2_controller_test.rb [hunk 5]
```diff
@@ -228,4 +249,70 @@ def teardown
...
+        assert_equal 0, comments(:second).flag
+    end
+      
+    test 'Admins should be able to unflag a comment' do
+        UserSession.create(users(:admin))
+        comment = comments(:second)
+        post :remove_flag_comment, params: { id: comment.id }
+        comment = assigns(:comment)
+        assert_equal 0, comments(:second).flag
+    end
```
Comment (jywarren) [inline test/functional/spam2_controller_test.rb:317]: These are great -- excellent for underlying functionality!

File: test/system/spam2_test.rb [hunk 6]
```diff
@@ -33,6 +33,13 @@ def setup
     page.assert_selector("div.alert", text: "Content spammed")
   end
 
+  test "unflag post in spam2" do
```
Comment (jywarren) [inline test/system/spam2_test.rb:36]: This is great too! Do you think it'd be possible to add a couple for regular users flagging posts, and for some other specific review action by moderators?

Comment (keshavsethi) [reply test/system/spam2_test.rb:36]: @jywarren I have added a system test where user can mark node as a flag. Please review Thanks!!

**Comment by keshavsethi** [comment]
> Hi @keshavsethi this is looking tremendous. I wonder if "Flag this node" should be "Flag this to moderators" as that way it's known who sees the flags, and the vocab word "node" isn't as confusing to folks?
> 
> Also, i wanted to see if you had any thoughts on how to complete #8122 because it occurred to me that it would make reviewing big PRs like this much easier; anyone could just boot it up in GitPod right from the PR! @Tlazypanda this could benefit some of your work as well. Would either of you be interested in helping to make progress on #8122 and that could speed up our review time quite a bit!

@jywarren That typo is done and  as #8122 is done Can we merge this. Thanks!!

**Comment by keshavsethi** [comment]
@jywarren @cesswairimu can we merge this ??

**Comment by cesswairimu** [comment]
Yeah looks good...we just need @jywarren final review against the changes he requested

**Comment by keshavsethi** [comment]
@jywarren can you please review and merge this as my future PRs are dependent on this. Please review 
Thanks!!

**Comment by keshavsethi** [comment]
@jywarren @cesswairimu you can test spam2 in gitpod Please review Thanks!!

**Comment by keshavsethi** [comment]
@jywarren @cesswairimu @pydevsg I have added the User moderation page here where the moderator can moderate users. I have also added filters for admin moderators and banned users. Can you Please review and merge this PR as I am stuck here. This PR is getting bigger and has so many changes. You can review it in gitpod. Thanks!!
Screenshot from 2020-07-20 10-52-26
In my future PRs, I am planning to add a search feature and **Queue** where the moderator can moderate only those nodes which have tags that they are following. 
Please review 
Thanks!!

**Comment by keshavsethi** [comment]
@jywarren @cesswairimu I have updated the UI a bit and completed user bulk moderation. Please review!!
This PR is getting Bigger, Please review and merge if it is fine. :pray: 
Screenshot from 2020-07-21 21-23-26
Please review and merge if it is fine. Thanks!!

**Comment by jywarren** [comment]
Reviewing this now! Thanks @keshavsethi for your patience. I hope we can work more modularly and that GitPod helps us review faster, so you get un-stuck!

**Comment by jywarren** [comment]
It's booting up in GitPod! 🕐 😄

**Comment by keshavsethi** [comment]
I have made some changes in gitpot https://github.com/publiclab/plots2/pull/8162 It is working fine there.

**Comment by jywarren** [comment]
OK!

Nice!

Screen Shot 2020-07-21 at 1 30 40 PM

**Comment by jywarren** [comment]
Hmm. i'm still getting hit by `Mysql2::Error: Expression #3 of SELECT list is not in GROUP BY clause` so i'm going to try to resolve that first! But the page looks great so i am pretty confident it'll be good to merge today!

File: test/system/dashboard_test.rb [hunk 7]
```diff
@@ -22,5 +22,16 @@ class DashboardTest < ApplicationSystemTestCase
     take_screenshot
 
   end
-
+  
+  test "User can flag a node from dashboard" do
```
Comment (jywarren) [inline test/system/dashboard_test.rb:26]: Awesome. Thanks!

**Comment by jywarren** [comment]
I'm content with how this looks so far! Thank you! The tests were super helpful and I loved being able to try it out on GitPod. Thanks and congratulations, @keshavsethi !!!

**Comment by keshavsethi** [comment]
Thank you so much @jywarren  :tada:  :tada: :heart:

**Pull request merged**